pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,293 @@
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package infracluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	mapiv1beta1 "github.com/openshift/api/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	cerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ptr "k8s.io/utils/ptr"
+	azurev1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	errUnableToGetAzureMAPIProviderSpec = errors.New("unable to get Azure MAPI ProviderSpec")
+	errUnableToGetAzureClientID         = errors.New("unable to get Azure Client ID")
+	errUnableToGetAzureTenantID         = errors.New("unable to get Azure Tenant ID")
+	errPlatformStatusNil                = errors.New("platform status should not be nil")
+)
+
+const (
+	clusterSecretName               = "capz-manager-cluster-credential"    // #nosec G101
+	capzManagerBootstrapCredentials = "capz-manager-bootstrap-credentials" // #nosec G101
+)
+
+// ensureAzureCluster ensures the AzureCluster cluster object exists.
+func (r *InfraClusterController) ensureAzureCluster(ctx context.Context, log logr.Logger) (client.Object, error) {
+	// Get the Azure Bootstrap Credentials Secret.
+	// This is created by the Cluster Credential Operator and should always exist. We expect to always find it, if not, we error.
+	capzManagerBootstrapSecret := &corev1.Secret{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: defaultCAPINamespace, Name: capzManagerBootstrapCredentials}, capzManagerBootstrapSecret); err != nil && !cerrors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get Azure Boostrap Credentials Secret: %w", err)
+	}
+
+	if err := r.ensureClusterSecret(ctx, *capzManagerBootstrapSecret); err != nil {
+		return nil, fmt.Errorf("error obtaining Azure Cluster Secret: %w", err)
+	}
+
+	if err := r.ensureClusterIdentity(ctx, *capzManagerBootstrapSecret); err != nil {
+		return nil, fmt.Errorf("error ensuring Azure Cluster Identity: %w", err)
+	}
+
+	target := &azurev1.AzureCluster{ObjectMeta: metav1.ObjectMeta{
+		Name:      r.Infra.Status.InfrastructureName,
+		Namespace: defaultCAPINamespace,
+	}}
+
+	log.Info(fmt.Sprintf("AzureCluster %s/%s does not exist, creating it", target.Namespace, target.Name))
+
+	if err := r.ensureAzureInfraCluster(ctx, target); err != nil {
+		return nil, fmt.Errorf("error ensuring Azure Infra Cluster: %w", err)
+	}
+
+	log.Info(fmt.Sprintf("InfraCluster '%s/%s' successfully created", defaultCAPINamespace, r.Infra.Status.InfrastructureName))
```

Comment:
This log line is in the wrong place, we should only print this if we do actually create the object, currently it will print even when the create is not achieved, perhaps move to line 247?

Reply:
Done!